### ISSUE TITLE
Manual Placement of Graph Nodes of High-level Optimizers and Loss functions

### ISSUE THREAD

**Issue by xilenteyex** [tensorflow/tensorflow#17713]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: 1.6.0-rc1
- **Python version**: 2.7
- **Bazel version (if compiling from source)**: 0.11.0
- **GCC/Compiler version (if compiling from source)**: 5.4
- **CUDA/cuDNN version**: 9.1/7.0
- **GPU model and memory**: Tesla k80 (11441MiB)
- **Exact command to reproduce**: python cifar10_train.py

### Describe the problem
I want to use my own device placement algorithm to be used to place nodes across multiple devices my machine has. I looked into this : https://www.tensorflow.org/programmers_guide/using_gpu#manual_device_placement which suggests that if I figure out exact placement statically before executing the graph, I can use "with device" annotations to partition my graph across devices. But, if I am using some high-level optimizers and loss functions. Each of these operations are going to have a multi-node graph as well. Suppose I am running this example : https://github.com/tensorflow/models/tree/master/tutorials/image/cifar10 and want to specify where each node of the complete tensor flow graph should go given a list of compute resources (cpu cores, gpu, tpu etc.) my machine has.  

Is there any api or framework that allows you to do fine grained placement of graph nodes across multiple compute resources ? What's the best way to achieve this ?  

Thanks a lot for any help!

### Source code / logs
cifar10 example : https://github.com/tensorflow/models/tree/master/tutorials/image/cifar10

**Comment by michaelisard** [comment]
@benoitsteiner is there documentation on whether Grappler could be used to plug in a third-party placement policy?

**Comment by facaiy** [comment]
Hi, I think stackoverflow could be a better place to ask the question. 

As far as I known, you can use device_function with `tf.device` to manage all ops placement by yourself, see the example below:

https://github.com/tensorflow/tensorflow/blob/6f0dd0425c51360fe2be5a938a8f3fb39e420fa3/tensorflow/python/framework/graph_util_test.py#L41-L45

However, in most case, `allow_soft_placement` is good enough, see [Using a single GPU on a multi-GPU system](https://www.tensorflow.org/programmers_guide/using_gpu#using_a_single_gpu_on_a_multi-gpu_system)

**Comment by xilenteyex** [comment]
Hi @facaiy ,
I want to run some heuristic after creating a complete graph and want to do fine grained placement of nodes of the graph while using high-level api's like loss functions, optimizers, functions from NN library etc. But using "with device annotations", I think I can only do placement on a very high-level. Please correct me if I am wrong. Thanks!

**Comment by facaiy** [comment]
You can pass a function to tf.device, and use the function to guide where each op is placed. see that `test_divice_function_ping_variable_to_cpu` in above example always assign all variable ops to cpu. 

with tf.device(your_placement_policy_function): 
 bulid your whole network

**Comment by xilenteyex** [comment]
Thanks, I will look into that.

**Comment by xilenteyex** [comment]
Hi @facaiy , what you suggested works fine in the case when you already know nodes -> device mapping before while creating the graph. But, for my case, I first want to construct the whole graph, now run some heuristic based algorithm on the complete graph and place them intelligently based on the nature of the complete graph. Do you know if this feature/flexibility already exists in Tensorflow ?

**Comment by facaiy** [comment]
@xilenteyex Sorry, I have no idea. 
cc @martinwicke Excuse me, do you know anyone who can answer the issue?

**Comment by Benyuel** [comment]
@xilenteyex: I think this is a basic example of what you're trying to do:

```
import tensorflow as tf
from tensorflow.python.framework import dtypes
from tensorflow.python.framework import meta_graph
from tensorflow.python.framework import ops as tf_ops
from tensorflow.python.ops import math_ops
from tensorflow.python.ops import nn_ops
from tensorflow.python.ops import random_ops
from tensorflow.python.ops import variable_scope

def _buildMnist(batch_size=128,
		input_size=256,
		num_classes=1024,
		num_layers=10,
		hidden_size=256,
		name='mnist'):
	g = tf_ops.get_default_graph()
	with g.as_default():
	  ops = {}
	  x = random_ops.random_uniform(
		  [batch_size, input_size], -0.1, 0.1, dtype=dtypes.float32)
	  for layer_id in range(num_layers):
		with variable_scope.variable_scope('layer_{}'.format(layer_id)):
		  a = input_size if layer_id == 0 else hidden_size
		  b = hidden_size if layer_id < num_layers - 1 else num_classes
		  w = variable_scope.get_variable('w', [a, b])
		  x = math_ops.matmul(x, w)
		  x = nn_ops.relu(x)
	  ops['y_preds'] = math_ops.argmax(x, axis=1)

	train_op = g.get_collection_ref(tf_ops.GraphKeys.TRAIN_OP)
	train_op.append(ops['y_preds'])
	return g

# build your graph
graph = _buildMnist()
# extract metagraph
mg = meta_graph.create_meta_graph_def(graph=graph)

# Assign your devices to nodes in the loop below
# As a simple example, assigning all ops to CPU.
# If you were to remove this, you would see some nodes assigned to GPUs by default
for node in metagraph.graph_def.node:
	node.device = '/device:CPU:0'
	
# define a new graph
new_graph = tf.Graph()
# log_device_placement so we can see the final placement
config = tf.ConfigProto(log_device_placement)
with tf.Session(graph=new_graph,
                config=config) as session:
	# load your modified metagraph
	saver = tf.train.import_meta_graph(mg)
	# initialize and run your train_op
	session.run(tf.global_variables_initializer())
	session.run(new_graph.get_collection_ref(tf_ops.GraphKeys.TRAIN_OP))
```

**Comment by xilenteyex** [comment]
@Benyuel thanks for your input. Your approach looks similar to what @facaiy suggested. I am more interested in an API/feature available in TensorFlow which allows you to do dynamic placement of nodes while executing the graph instead of statically assigning nodes to devices before graph execution.  For example, If I have 10 nodes in the graph, after executing first three nodes based on the static placement, is it possible to change the placement of next few nodes or all nodes that are going to be executed ?

**Comment by Benyuel** [comment]
You might be able to adapt something from how eager execution works: https://www.tensorflow.org/programmers_guide/eager

**Comment by tensorflowbutler** [comment]
Nagging Assignee @benoitsteiner: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
